universe/ValueRefs.h

```diff
@@ -760,6 +789,95 @@ void Variable<T>::serialize(Archive& ar, const unsigned int version)
         & BOOST_SERIALIZATION_NVP(m_return_immediate_value);
 }
 
+///////////////////////////////////////////////////////////
+// NamedRef                                              //
+///////////////////////////////////////////////////////////
+template <typename T>
+NamedRef<T>::NamedRef(const std::string& value_ref_name) :
+    m_value_ref_name(value_ref_name)
+{
+    DebugLogger() << "ctor(NamedRef<T>): " << typeid(*this).name() << " value_ref_name: " << m_value_ref_name;
+}
+
+template <typename T>
+bool NamedRef<T>::operator==(const ValueRef<T>& rhs) const
+{
+    if (&rhs == this)
+        return true;
+    if (typeid(rhs) != typeid(*this))
+        return false;
+    const NamedRef<T>& rhs_ = static_cast<const NamedRef<T>&>(rhs);
+    return (m_value_ref_name == rhs_.m_value_ref_name);
+}
+
+template <typename T>
+bool NamedRef<T>::RootCandidateInvariant() const
+{ return GetValueRef() ? GetValueRef()->RootCandidateInvariant() : true; }
+
+template <typename T>
+bool NamedRef<T>::LocalCandidateInvariant() const
+{ return GetValueRef() ? GetValueRef()->LocalCandidateInvariant() : true; }
+
+template <typename T>
+bool NamedRef<T>::TargetInvariant() const
+{ return GetValueRef() ? GetValueRef()->TargetInvariant() : true; }
+
+template <typename T>
+bool NamedRef<T>::SourceInvariant() const
+{ return GetValueRef() ? GetValueRef()->SourceInvariant() : true; }
+
+template <typename T>
+std::string NamedRef<T>::Description() const
+{ return GetValueRef() ? GetValueRef()->Description() : std::string("NAMED_REF_UNKNOWN"); }
```

Comment:
should return `UserString` lookup results.

Reply:
should i use the UserString also for ::Dump ?

Reply:
No.

`Dump` should ideally reproduce a FOCS script that could be parsed back into the same data structure, or as close to it as practical.

`Description` should be a user-readable / localized bit of text that says what the object does.